src/irmin-pack/checks.ml

```diff
@@ -363,3 +421,167 @@ module Index (Index : Pack_index.S) = struct
     Utils.Progress.finalise bar;
     result
 end
+
+module Stats (S : sig
+  type step
+
+  val step_t : step Irmin.Type.t
+
+  module Hash : Irmin.Hash.S
+end) =
+struct
+  type path = S.step option list
+  type paths = S.step option list list
+
+  type stat = {
+    all_paths : paths;
+    (* All path to an object. *)
+    max_length : int;
+    (* The max length of a path to an object. *)
+    mp : int * int * path;
+        (* The max mp, the nb of paths with max mp and one example. *)
+  }
+
+  type t = {
+    visited : (S.Hash.t, stat) Hashtbl.t;
+    mutable max_width : int * int * path;
+    (* The max, the number of nodes with the max and one example. *)
```

Comment:
There's maybe also potential for defining this in its own module and providing `empty` / `add` / `pp` (like https://github.com/mirage/irmin/blob/master/src/irmin-pack/reconstruct_index.ml#L4-L33), to avoid interleaving the update logic with the traversal logic, but it's just personal preference

Reply:
Thanks for these suggestions! I introduced an internal module Metrics, that handles the updates. I agree that the code is a bit more readable now.